File: go/promotions/promotion.go
// ABOUTME: Promotion interface defining the contract for promotions.
// ABOUTME: Promotions can check applicability and apply discounts to carts.
package promotions

import (
	"promotionengine/domain"
	"promotionengine/pricing"
)

// Promotion represents a promotional discount that can be applied to a cart.
// TODO: Implement specific promotion types (e.g., PercentOffdomain.Cart, BuyXGetY, etc.)
type Promotion interface {
	// ID returns the unique identifier for this promotion.
	ID() domain.PromotionID

	// IsApplicable checks if this promotion can be applied to the given cart and context.
	IsApplicable(cart domain.Cart, context pricing.PricingContext) bool

	// Apply applies this promotion to the cart and returns the resulting discounts.
	Apply(cart domain.Cart, context pricing.PricingContext) []pricing.AppliedDiscount
}

File: typescript/src/promotions/Promotion.ts
// ABOUTME: Interface defining the contract for promotion implementations
// ABOUTME: Promotions evaluate cart eligibility and generate discount applications

import { Cart } from '../domain/Cart';
import { PromotionId } from '../domain/PromotionId';
import { PricingContext, AppliedDiscount } from '../pricing';

// TODO: Add time-based applicability (validFrom, validUntil)
// TODO: Add day-of-week and time-of-day restrictions
// TODO: Add customer segment targeting using context.customerTags
// TODO: Implement specific promotion types:
//   - PercentOffProduct: Apply percentage discount to specific products
//   - BuyXGetY: Buy X items, get Y items free/discounted
//   - FixedAmountOff: Fixed dollar amount off
//   - ThresholdPromotion: Minimum purchase amount required

export interface Promotion {
  id(): PromotionId;

  isApplicable(cart: Cart, context: PricingContext): boolean;

  apply(cart: Cart, context: PricingContext): AppliedDiscount[];
}
